app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/ToolbarUiController.kt

```diff
@@ -36,17 +45,40 @@ class ToolbarUiController(
 
     private var hasUserChangedURLSinceEditTextFocused = false
 
+    private lateinit var tooltip: PopupWindow
+    private lateinit var tooltipView: View
+
     fun onCreateView(layout: View, viewLifecycleOwner: LifecycleOwner, fragmentManager: FragmentManager) {
         val toolbarClickListener = ToolbarOnClickListener()
         layout.topNavContainer.forEach {
             it.nextFocusDownId = layout.navUrlInput.id
             if (it.isFocusable) it.setOnClickListener(toolbarClickListener)
+
+            it.setOnFocusChangeListener { _, hasFocus ->
+                if (hasFocus) showTooltip(it)
+                else tooltip.dismiss() // Hide the tooltip when the button is not focused
+            }
         }
 
+        val layoutInflater = layout.context.systemService<LayoutInflater>(Context.LAYOUT_INFLATER_SERVICE)
+        tooltipView = layoutInflater.inflate(R.layout.tooltip, null)
+        tooltip = PopupWindow(tooltipView, WRAP_CONTENT, WRAP_CONTENT, false)
+
         setupUrlInput(layout)
         initToolbar(layout, viewLifecycleOwner, fragmentManager)
     }
 
+    private fun showTooltip(navBarButton: View) {
+        tooltip.contentView.tooltip.text = navBarButton.contentDescription
+        tooltip.isClippingEnabled = false
+
+        // Since the height and width are WRAP_CONTENT, we need to force the layout measure to happen before we can
+        // access the true dimensions
```

Comment:
Glad you found a solution, and this makes sense why this wasn't working before.

I looked into this more, because using `measure()` directly is usually a red flag:
By calling `measure` here means it will get called twice - once here, and once during the normal layout pass that the View does (wherever this would actually get measured). Measuring is usually expensive (because it has to recurse through all the children) but in this case, we *know* that it's only a single TextView, so it's not *terrible* - if this view got more complicated though, we'd want to change this to do it correctly.

The correct way to do this would be to hook into the normal layout pass by setting an OnGlobalLayoutListener, which will get called during the normal layout pass, and then call `tooltip.update` with the new offset, because the height will be guaranteed to have been set at that time. This way, we only call `measure` once.

A few options here:
a) Do it the correct way, tradeoff is that this way is "okay" and that will add a little more dev time, but it makes our code base cleaner/correct (e.g., in the case someone new comes in, they wouldn't be able to copy this antipattern somewhere else)
b) Add a comment explaining why this is okay in this one case (single TextView), and optionally include an explanation of the correct way. Say this is *NOT* how it should be done elsewhere.